keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,428 @@
+1. `bucketAccessClassName`: name of the `BucketAccessClass` specifying the desired set of policy actions to be set for a user identity or ServiceAccount.
+1. `bucketAccessName`: name of the bound cluster-scoped `BucketAccess` instance
+1. `message`: a human readable description detailing the reason for the current `phase``.
+1. `phase`: is the current state of the Bucket:
+   - _Bound_: the controller finished processing and has bound the BucketAccess to the BucketRequest via the BucketRequest.Spec.BucketAccessRequestName and BucketRequest.Spec.BucketAccessRequestNamespace
+   - _Failed_: error and all retries have been exhausted.
+   - _Retrying_: set when a driver or Kubernetes error is encountered during provisioning operations indicating a retry loop.
+
+#### BucketAccess
+
+A cluster-scoped administrative API which encapsulates fields from the BucketAccessRequest and the BucketAccessClass.  The purpose of the API is to server as communication path between provisioner and a central COSI controller.  In Greenfield, the COSI controller will create BucketAccess instances for new BucketAccessRequests.  The relationship between a BucketAccessRequests and a BucketAccess should be 1:1.  
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketAccess
+metadata: 
+  name: [1]
+  labels:
+    cosi.io/provisioner: [2]
+  finalizers:
+  - cosi.io/finalizer [3]
+ spec:
+  bucketAccessRequestName: [4]
+  bucketAccessRequestNamespace: [4]
+  serviceAccountName: [5]
+  accessSecretName: [6]
+  provisioner: [7]
+  parameters: [8]
+ status:
+  message: [9]
+  phase: [10]
+```
+
+1. `name`: For Greenfield, generated in the pattern of `bucketAccessRequest-<my-bucketAccessRequest>-<my-bucketAccessRequest-namespace>`. 
+1. `labels`: should be added by controller.  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
+1. `finalizers`: should be added by controller to defer `BucketAccess` deletion until backend deletion ops succeed.
+1. `bucketAccessRequestName`/`bucketAccessRequestNamespace`: name and namespace of the bound `BucketAccessRequest`
+1. `serviceAccountName`: name of the Kubernetes ServiceAccount specified by the `BucketAccessRequest`.  Undefined when the `BucketAccessRequest.accessSecretName` is defined.
+1. `  accessSecretName`: name of the *provisioner* generated `Secret` containing access credentials. This `Secret` exists in the provisioner’s namespace and must be copied to the app namespace by the COSI controller.
+1. `provisioner`:  name of the provisioner that should handle this `BucketAccess` instance.  Copied from the `BucketAccessClass`.
+1. `parameters`:  A map of string, string key values.  Allows admins to control user and access provisioning by setting provisioner key-values.
```

Comment:
@wlan0 is this different from `bucketAccessClass.parameters`?

Reply:
This is an additonal set of parameters on top of what is available in the class. An example use case would be where the bucketClass has a set of defaults and this field allows further customization on top of that.

Reply:
> where the bucketClass has a set of defaults and this field allows further customization on top of that.

@wlan0 1) do you mean BucketClass or BucketAccessClass above?
2) I do not understand if there are any differences between green- and brownfield for BARs, BACs, and BAs like there are with bucket provisioning.
3) I assume that a BA is created manually by the admin who also creates the BAC.

Reply:
> do you mean BucketClass or BucketAccessClass above? 

yes, it should've been bucketAccessClass above

> 3)  I assume that a BA is created manually by the admin who also creates the BAC.

This is one possibility. It is also possible that the BA for brownfield is autogenerated, for instance, in the Greenfield -> brownfield case, each bucket accessRequest gets a new set of credentials for that workload